### DISCUSSION TITLE
"Official" way to use third-party tools?

### DISCUSSION THREAD

**Discussion by jones-michael-s** [automaticanalysis/automaticanalysis#253] [Q&A] [answered]
Hello.

I'm getting ready to open a PR for our denoising tools. 

Question: My code uses rWLS, Wavelet toolbox (WDS), and AROMA.

I downloaded rWLS and WDS someplace convenient on my machine and addpath to them.

My AROMA module assumes AROMA lives in aap.directory_conventions_fsldir/bin/ICA-AROMA.

It all works, but are these acceptable solutions?  Or is there an official way to install third-party stuff for use with aa? 

Thanks much,
Mike

**Comment by tiborauer** [comment]
Since ICA-AROMA seems to be a binary and part of FSL, you do not really have to integrate (i.e. add to MATLAB path). However, you should use `checkrequirement` of the corresponding module to ensure ICA-AROMA is installed.

rWLS and WDS, on the other hand, requires handling from within MATLAB (add to MATLAB path); therefore, they should be integrated via the toolbox framework (each toolbox separately):

1. You need to write interface (<toolbox name>Class.m). See the [toolboxes folder](https://github.com/automaticanalysis/automaticanalysis/tree/master/aa_tools/toolboxes) for inspiration.
2. The corresponding module should `check`, `load`, and `unload` the toolbox. See [aamod_tdt_decode (lines 28-29, 188, 191)](https://github.com/automaticanalysis/automaticanalysis/blob/master/aa_modules/aamod_tdt_decode.m) as an example.
3. Your (site-specific) parameterset should contain a corresponding toolbox entry. See [aap_parameters_defaults_UoS.xml (lines 21-57)](https://github.com/automaticanalysis/automaticanalysis/blob/master/aa_parametersets/aap_parameters_defaults_UoS.xml)

**Comment by jones-michael-s** [comment]
Counter-offer: I PR my code as-is and you modify it to conform to the new interface. That would really help us out on this.

I suspect this would take you about a half hour. It would take me a month.

**Comment by tiborauer** [comment]
You think a bit too highly of me :), and I am afraid I will not have time for that. I also do not think it is a sustainable development model.

The question is whether you consider the toolbox framework a good solution. If yes, then it may be worth your time to invest now. If not, you may need to propose an alternative solution.

**Comment by jones-michael-s** [comment]
The main difficulty is I can find no documentation on writing an aa class interface. Is there any? Perhaps I am looking in the wrong place.

Since you have made the toolbox framework a prerequisite for aa development, you are obligated to provide such documentation. 
For example, when I rewrote the testing utility, I provided an extensively commented example script (aatest_ds000114_TEMPLATE) that explains in detail how to write a testscript when a developer contributes new code. In contrast, there is no such example of a documented toolbox interface. What is required? What is optional? What do the options mean? How are they used? Where are they used? When are they used? There is no explanation whatsoever .In fact, I can't find even one comment in any of the examples in aa_tools/toolboxes.. 

The alternative is spending a month stepping though the toolbox code trying to figure out what it's supposed to do. I also do not think this is a sustainable development model.

**Comment by tiborauer** [comment]
You are absolutely right! We need proper documentation, and not only for the toolbox framework. 
And I wish I had that that extra bandwidth you had for writing that extensive documentation (something similar should also go to the website, as well). However, I am not a professional developer, and I am not paid for the development of aa; therefor I feel "obligated" is a bit strong expectation. On the other hand, I do feel somewhat obligation to other fellow developers (that is you, mostly), if I expect them to build upon what I did; but I am not sure whether I am always be able to fulfil it. This is one of the main issues of the extracurricular/ open-source development. 
When I took over aa, I had even less documentation, so I am used to "spending a month stepping though the code trying to figure out what it's supposed to do." And often I did not have access to the original developer(s), or even they did not know. :) And it is not a sustainable development model, either, indeed, but what could I do? So forgive me, if I am used to that. I thought those [three short notes](https://github.com/automaticanalysis/automaticanalysis/discussions/253#discussioncomment-1390605) will be enough to go ahead, but I understand that they are not as extensive as your documentation.

This discussion goes beyond the current issue, and I am not even sure whether there is any answer for that apart from dedicated funding for open-source development. More importantly, I certainly do not have the bandwidth to provide documentation for everything I ever did with aa. Maybe if I had a different job. :)

You asked me about the "official" way to use third-party tools, and my "two cent" is the toolbox interface - even though it is still under construction (esp. in terms of documentation). Apart from (or before) that, there is/was nothing, which I strongly oppose. To provide an answer to your issue, at least, you either hold your PR until the toolbox interface is properly documented so that you feel confident implementing it, or you submit your PR now and extend it later. However, I would like to avoid "officially" encouraging different practices.

**Comment by jones-michael-s** [comment]
I don't disagree with anything you wrote, but it still leaves me with no solution. Our lab just finished a study that uses my new code, we can't submit our paper until my PR goes in, and my PR can't go in until rWLS and WDS get a class interface.

I'll run this past the boss and see what he wants to do.

**Comment by tiborauer** [comment]
There are multiple solutions:
1. You can cite your repository.
2. You can create a new branch in the main repository and merge your PR into it. Later, it can be merged into the `master` branch.

**Comment by tiborauer** [comment]
I have added some documentation for the toolbox framework to the [website](https://automaticanalysis.github.io/_pages/toolboxes.html) and updated the [core concepts](https://automaticanalysis.github.io/about/#core-concepts) accordingly.

The same documentation will also be added to the [toolboxes](https://github.com/automaticanalysis/automaticanalysis/tree/master/aa_tools/toolboxes) to keep it close to the code.

Please, take a look and let me know whether you find it instructive enough.

N.B.: It will also require a corresponding PR from me to work. I will submit it soon.

**Comment by jones-michael-s** [comment]
Thanks much. I'll have a look...

**Comment by jones-michael-s** [comment]
Quick question: I'm looking at wavelet despiking (it's one of the few SPM toolboxes I'm familiar with).

aamod_waveletdespike.m has a call to aas_cache_get at the top, suggesting the module has been modified to conform to the new toolbox interface. However, I don't see a class interface file for wavelet despiking in aa_tools/toolboxes. 

Is this an omission or are there some toolboxes that don't require a class interface?

**Reply by tiborauer** [reply]
The aas_cache_get loads the SPM toolbox and not the BrainWavelet Toolbox, which has not been "toolboxed" yet.

The "old" practice has been to either dump the path(s) to `aap.directory_conventions.matlabtoolsdir` or add a dedicated setting, such as  `aap.directory_conventions.BrainWaveletdir`. My intention is to integrate all external toolboxes via an interface.

**Comment by jones-michael-s** [comment, marked as answer]
Ok, I have "toolboxerized" the BrainWavelet toolbox. My updated code runs without crashing but let me list the changes I made here and you can let me know if I did anything horrific.

1) I added wdsClass.m -- I called the class wds (waveletdespiking) and it's mostly just fwsClass with a name change and a addpath(genpath()) to pick up nested mfiles. (I'll include the code listing at the end so it makes my list easier to read)

2) I load the toolbox on entry to aamod_waveletdespike.m (which also loads spm) and unload on exit:

```
    case 'doit'        
        [ ~, SPMtool ] = aas_cache_get(aap,'spm');        
        [ ~, WDS ] = aas_cache_get(aap,'wds');
        WDS.load;
```


```
        %% Describe outputs
        aap=aas_desc_outputs(aap,aap.tasklist.currenttask.domain,[subj sess],streams{1},outputfns);
        
        %% unload
        WDS.unload;
```


3a) I added a wds toolbox entry to my parameter file:

```
        <toolbox desc='Toolbox with implemented interface in extrafunctions/toolboxes' ui='custom'>
            <name desc='Name corresponding to the name of the interface without the "Class" suffix' ui='text'>wds</name>
            <dir ui='dir'>/path/to/my/BrainWavelet</dir>
            <extraparameters>
                <doKeepInPath>1</doKeepInPath>
            </extraparameters>
        </toolbox>
```

3b) I removed the old BrainWavelet toolbox parameter file entry:

`<BrainWaveletdir desc='Path to BrainWavelet toolbox' ui='dir'>/path/to/my/BrainWavelet</BrainWaveletdir>    DELETED
`

4) I deleted the path setup code for BrainWavelet currently in aa_init:

lines 146-158 DELETED:

```
% Path to BrainWavelet
if ~isempty(aap.directory_conventions.BrainWaveletdir)
    addpath(fullfile(aap.directory_conventions.BrainWaveletdir,'BWT'));
    addpath(fullfile(aap.directory_conventions.BrainWaveletdir,'third_party','cprintf'));
    addpath(fullfile(aap.directory_conventions.BrainWaveletdir,'third_party','NIfTI'));
    addpath(fullfile(aap.directory_conventions.BrainWaveletdir,'third_party','wmtsa','dwt'));
    addpath(fullfile(aap.directory_conventions.BrainWaveletdir,'third_party','wmtsa','utils'));
else
    % Check whether already in path, give warning if not
    if isempty(which('WaveletDespike'))
       aas_log(aap,false,sprintf('BrainWavelet not found, if you need this you should add it to the matlab path manually, or set aap.directory_conventions.BrainWaveletdir'));
    end
end
```

and lines 232-238 DELETED:

```
% BrainWavelet
if ~isempty(aap.directory_conventions.BrainWaveletdir)
    p_ind = cell_index(p,aap.directory_conventions.BrainWaveletdir);
    for ip = p_ind
        reqpath{end+1} = p{ip};
    end
end
```

I think that's all.

Here's the complete classDef file listing:

```
% aa toolbox interface for BrainWavelet toolbox

classdef wdsClass < toolboxClass
    
    properties (Access = protected)
        hGUI = []% GUI handles
    end
    
    methods
        
        function obj = wdsClass(path,varargin)
            
            defaultAddToPath = false;
            defaultKeepInPath = false;
            
            argParse = inputParser;
            argParse.addRequired('path', @ischar);
            argParse.addParameter('name','', @ischar);
            argParse.addParameter('doAddToPath', defaultAddToPath,@(x) islogical(x) || isnumeric(x));
            argParse.addParameter('doKeepInPath', defaultKeepInPath,@(x) islogical(x) || isnumeric(x));
            argParse.parse(path,varargin{:});
            
            obj = obj@toolboxClass(argParse.Results.name,argParse.Results.path,argParse.Results.doAddToPath,argParse.Results.doKeepInPath,{});
        end
        
        function load(obj)
            addpath(genpath(obj.toolPath));

            load@toolboxClass(obj)
        end
        
    end
    
end
```

**Reply by tiborauer** [reply]
One small comment: you probably do not want to add 
```xml
    <extraparameters>
        <doKeepInPath>1</doKeepInPath>
    </extraparameters>
```
to the parameterset. It was (mostly) reserved for SPM and has no effect anywhere. 

N.B.: It will be also phased out by my #255 along with the corresponding argumentation in the interfaces.

**Comment by jones-michael-s** [comment]
Matlab appears to require an extraparameters field appear in each toolbox entry, otherwise aa_init crashes on this:

```
if isfield(aap.directory_conventions,'toolbox') && isstruct(aap.directory_conventions.toolbox)
    for TBX = aap.directory_conventions.toolbox
        if strcmp(TBX.name,'spm'), continue; end
        aas_cache_put(aap,TBX.name,aas_inittoolbox(aap,TBX.name));
    end
end
```

Is there something you would recommend to replace doKeepInPath?

**Comment by tiborauer** [comment]
I find it odd, however, you can keep the fied empty, as well:
````xml
    <extraparameters></extraparameters>
````

It is also possible that this bug has been fixed by #255 only.
